packages/evm/contracts/core/modules/ACL.sol

```diff
@@ -0,0 +1,96 @@
+
+    enum BulkOp { Grant, Revoke, Freeze }
+
+    address internal constant FREEZE_FLAG = address(1);
+    address internal constant ANY_ADDR = address(-1);
+
+    // List of all roles assigned to different addresses
+    mapping (bytes32 => mapping (address => bool)) public roles;
+
+    event Granted(bytes32 indexed id, address indexed who);
+    event Revoked(bytes32 indexed id, address indexed who);
+    event Frozen(bytes32 indexed id);
+
+    modifier notFrozen(bytes32 _id) {
+        require(!isRoleFrozen(_id), ERROR_ROLE_ALREADY_FROZEN);
+        _;
+    }
+
+    /**
+    * @dev Tell whether an address has a role assigned
+    * @param _who Address being queried
+    * @param _id ID of the role being checked
+    * @return True if the requested address has assigned the given role, false otherwise
+    */
+    function hasRole(address _who, bytes32 _id) public view returns (bool) {
+        return roles[_id][_who] || roles[_id][ANY_ADDR];
+    }
+
+    /**
+    * @dev Tell whether a role is frozen
+    * @param _id ID of the role being checked
+    * @return True if the given role is frozen, false otherwise
+    */
+    function isRoleFrozen(bytes32 _id) public view returns (bool) {
+        return roles[_id][FREEZE_FLAG];
+    }
+
+    /**
+    * @dev Internal function to grant a role to a given address
+    * @param _id ID of the role to be granted
+    * @param _who Address to grant the role to
+    */
+    function _grant(bytes32 _id, address _who) internal notFrozen(_id) {
+        require(_who != FREEZE_FLAG, ERROR_BAD_FREEZE);
+        require(!hasRole(_who, _id), ERROR_ROLE_GRANTED);
+        roles[_id][_who] = true;
+        emit Granted(_id, _who);
+    }
+
+    /**
+    * @dev Internal function to revoke a role from a given address
+    * @param _id ID of the role to be revoked
+    * @param _who Address to revoke the role from
+    */
+    function _revoke(bytes32 _id, address _who) internal notFrozen(_id) {
+        require(hasRole(_who, _id), ERROR_ROLE_NOT_GRANTED);
+        roles[_id][_who] = false;
+        emit Revoked(_id, _who);
+    }
+
+    /**
+    * @dev Internal function to freeze a role
+    * @param _id ID of the role to be frozen
+    */
+    function _freeze(bytes32 _id) internal notFrozen(_id) {
+        roles[_id][FREEZE_FLAG] = true;
+        emit Frozen(_id);
+    }
+
+    /**
+    * @dev Internal function to bulk a set of ACL operations
+    */
+    function _bulk(BulkOp[] memory _op, bytes32[] memory _id, address[] memory _who) internal {
+        require(_op.length == _id.length && _op.length == _who.length, ERROR_INVALID_BULK_INPUT);
+
+        for (uint256 i = 0; i < _op.length; i++) {
+            BulkOp op = _op[i];
+            if (op == BulkOp.Grant) {
+                _grant(_id[i], _who[i]);
+            } else if (op == BulkOp.Revoke) {
+                _revoke(_id[i], _who[i]);
+            } else if (op == BulkOp.Freeze) {
+                _freeze(_id[i]);
```

Comment:
I guess in this case, you would put in anything you wanted for the `_who` element (most likely the zero address to save calldata gas)?

Reply:
Yup, exactly